src/braket/circuits/synthesis/one_qubit_decomposition.py

```diff
@@ -0,0 +1,211 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+
+import numpy as np
+import braket.circuits as braket_circ
+
+from braket.circuits.synthesis.util import to_su
+from braket.circuits.quantum_operator_helpers import is_unitary
+from braket.circuits import Circuit
+from braket.circuits.gates import X, Y, Z, Rx, Ry, Rz
+from braket.circuits.instruction import Instruction
+
+y = Y().to_matrix()
+z = Z().to_matrix()
+
+
+class OneQubitDecomposition:
+    """
+    Class for decomposition of 1-qubit gates.
+
+    Args:
+        U (np.ndarray): the unitary matrix to decompose.
+
+    Attributes:
+        phase (float): the global phase.
+        canonical_vector (np.ndarray): the axis about which the rotation is
+            performed.
+        rotation_angle (float): the amplitude of the rotation.
+        quaternion (np.ndarray): the quaternion representation
+            of the rotation.
+        rotation_matrix (np.ndarray): the matrix representation
+            of the rotation.
+    """
+
+    def __init__(self, U: np.ndarray, atol: float = 1e-8, rtol: float = 1e-5):
+
+        self.atol = atol
+        self.rtol = rtol
+
+        if U.shape != (2, 2):
+            raise ValueError("Input matrix has to be 2x2.")
+
+        is_unitary(U, atol=self.atol, rtol=self.rtol, raise_exception=True)
+
+        self.U = U
+        su = to_su(self.U)
+        self.phase = complex(np.linalg.det(U)) ** 0.5
+
+        # Calculate zyz Euler angles.
+        weight_matrix = np.array([[1, 1], [-1, 1]])
+
+        r1 = 2 * np.arctan2(abs(su[1, 0]), abs(su[0, 0]))
+
+        r0r2 = np.angle(su[1, :])
+
+        # have to check if r2 is 0 to avoid gimbal lock.
+        if abs(su[0, 1]) < self.atol:
+            r0 = 2 * np.angle(su[1, 1])
+            r2 = 0
+        else:
+            r0r2 = weight_matrix @ np.angle(su[1, :])
+            r0, r2 = np.where(r0r2 < 0, r0r2 + 2 * np.pi, r0r2)
+
+        self._euler = np.array([r0, r1, r2])
+
+        # Calculate the quaternion representation.
+
+        q0 = np.real(su[0, 0])
+        q1 = -np.imag(su[0, 1])
+        q2 = -np.real(su[0, 1])
+        q3 = -np.imag(su[0, 0])
+
+        self._quat = np.array([q0, q1, q2, q3])
+
+    def euler_angles(self, axes: str = "zyz"):
```

Comment:
Let's just use the zyz convention; this is standard in quantum mechanics, and it only complicates things to have a convention option at this level. We can introduce the classical xyz convention if it gets requested.